apps/lib/apps.c

```diff
@@ -1008,13 +1012,16 @@ int load_key_certs_crls(const char *uri, int format, int maybe_stdin,
             BIO_printf(bio_err, "Could not open file or uri for loading");
         goto end;
     }
+    /* expect == 0 means here multiple types of credentials are to be loaded */
     if (expect > 0 && !OSSL_STORE_expect(ctx, expect)) {
```

Comment:
Is `expect > 0` possible here???

Reply:
Yes - but the new comment is just about the `expect == 0` case.

Reply:
@vdukhovni That should be a separate PR then.

Reply:
I disagree, the code is rather weird as-is, the `expect` value is definitely zero after the first two calls to `SET_EXPECT1` on lines 958 and 959, whose sole effect is then to set the relevant pointers to `NULL`, and all the subsequent logic around `expect` is then either wrong or unnecessary.  It is hard to say that the code is correct in its current state.

What's going on on lines 968 and 977? It is appropriate at this point to have a clear idea of whether this function is ever reading just one type of object `expect != 0`, or whether it is always loading more than one type, and if so, all this `expect` stuff is bogus.

Reply:
> I disagree, the code is rather weird as-is, the `expect` value is definitely zero after the first two calls to `SET_EXPECT1` on lines 958 and 959, whose sole effect is then to set the relevant pointers to `NULL`, and all the subsequent logic around `expect` is then either wrong or unnecessary. It is hard to say that the code is correct in its current state.

No - `expect > 0` is even the usual case, 
namely that just one type of credential (privkey, pubkey, params, cert, crl) is expected!
See also the comment
```
/* Set type expectation, but set to 0 if objects of multiple types expected. */
```
which I just adapted slightly for more clarity.

Note the guard `(pvar) != NULL` in the `SET_EXPECT1(pvar, val)` macro.
Lines 958 to 961 have the following effect, which I just added as a comment:
```
     * If just one of the ppkey, ppubkey, pparams, and pcert function parameters
     * is nonzero, expect > 0 indicates which type of credential is expected.
     * If expect == 0, more than one of them is nonzero (multiple types expected).
```

> What's going on on lines 968 and 977? It is appropriate at this point to have a clear idea of whether this function is ever reading just one type of object `expect != 0`, or whether it is always loading more than one type, and if so, all this `expect` stuff is bogus.

I've also added respective comments:
```
         * Adapt the 'expect' variable:
         * set to OSSL_STORE_INFO_CERT if no other type is expected so far,
         * otherwise set to 0 (indicating that multiple types are expected).
```
```
         * Adapt the 'expect' variable:
         * set to OSSL_STORE_INFO_CRL if no other type is expected so far,
         * otherwise set to 0 (indicating that multiple types are expected).
```